### PR TITLE
Adding x-b3-* headers parsing and propagating support

### PR DESCRIPTION
**Pull request by kunalekawde**
## Which problem is this PR solving?
- #125 and probably  #141

## Short description of the changes
- The changes are parse / decode the tracing information received in x-b3-traceId, x-b3-spanId, x-b3-parentspanId, x-b3-sampled and to propagate them further.

### CODE REVIEW

File: src/jaegertracing/SpanContext.h [hunk 1]
```diff
@@ -45,6 +45,7 @@ class SpanContext : public opentracing::SpanContext {
         , _spanID(0)
         , _parentID(0)
         , _flags(0)
+        , _b3Headers(false)
```
Comment (yurishkuro) [inline src/jaegertracing/SpanContext.h]: This doesn't belong here. Parsing of B3 headers is the job for codecs (see https://github.com/jaegertracing/jaeger-client-cpp/tree/master/src/jaegertracing/propagation). You can use Jaeger clients in other languages as a reference.

**Comment by kunalekawde** [comment]
> 
> 
> You can use Jaeger clients in other languages as a reference.

Sure thanks, shall try that

**Comment by kunalekawde** [comment]
@yurishkuro Needed some help
1. Can we assume that uber-ctx-* & x-b3-* would be mutually exclusive to any node ?
2. As I see in java implementation (https://github.com/jaegertracing/jaeger-client-java/tree/master/jaeger-core/src/main/java/io/jaegertracing/internal/propagation) the codecs are interpreted as text, http, binary and b3. If we have to same at jaeger cpp client wont it be a substantial change w.r.t class design ? Moreover B3 is an alternative to Uber-Ctx, does it fit to be in classification of codes inline to text/http/binary ? Wouldn't B3 also ever need binary/text ?. 
Please let me know your comments which would help me further.

**Comment by yurishkuro** [comment]
1 - yes, I would treat them as mutually exclusive. If someone really wants to support both simultaneously, they can implement another delegating codec on top of those two. 

2- text, http, and binary are the three formats defined in opentracing. B3 would only implement text and http, probably with the same codec since b3 syntax is http-safe (unlike Jaeger’s baggage that requires url-encoding for values in http).

**Comment by kunalekawde** [comment]
> 2- text, http, and binary are the three formats defined in opentracing. B3 would only implement text and http, probably with the same codec since b3 syntax is http-safe (unlike Jaeger’s baggage that requires url-encoding for values in http).
 So can we have B3 at level of Uber-ctx type - probably abstracting common layer and then provide text/http/other encoding to it. Basically unlike java implementation then B3 wont be a codec which implements extract and inject but a user of it.

**Comment by yurishkuro** [comment]
If there's a common layer to abstract, I'm for it, but in the other clients I didn't see that, as the core of the codecs is dealing with specific headers of a given context format.

File: src/jaegertracing/SpanContext.h [hunk 2]
```diff
@@ -45,6 +52,7 @@ class SpanContext : public opentracing::SpanContext {
         , _spanID(0)
         , _parentID(0)
         , _flags(0)
+        , _tracerType(TracerType::TRACER_TYPE_UNKNOWN)
```
Comment (yurishkuro) [inline src/jaegertracing/SpanContext.h]: Why is this needed? The usual approach is to configure tracer with one type of codecs, which apply to both inbound and outbound requests. There’s no need to remember what type of encoding was used in the inbound.

Comment (kunalekawde) [reply src/jaegertracing/SpanContext.h]: But how and where do we need to configure a particular type of codec and based on what?
Here in this PR, I've based it on the incoming request headers. As it would be either B3 / Uber, one of the tracer type would be set. If the trace is initiated by that node itself, it would be set with options in startSpanWithOptions API. 
If we dont remember type of tracing then using injectSpan, we need to have some flag to indicate what type of tracing should be done so that appropriate headers can be set.

Comment (yurishkuro) [reply src/jaegertracing/SpanContext.h]: See https://github.com/jaegertracing/jaeger-client-java/pull/694/files, specifically `CodecConfiguration.withPropagation`

Comment (kunalekawde) [reply src/jaegertracing/SpanContext.h]: @yurishkuro thanks for the reference, I still fail to understand on how this implementation can be used dynamically. As per readme:
```
b3Codec = new B3TextMapCodec();
tracer = new JaegerTracer.Builder(serviceName)
  .registerInjector(Format.Builtin.HTTP_HEADERS, b3Codec)
  .registerExtractor(Format.Builtin.HTTP_HEADERS, b3Codec)
  ...
```
I also checked in stagemonitor.git, looks like application is/have to build a object at start by specifying the tracing type e.g. B3 / Uber. 

If this is so then, application is aware what kind of headers it is expecting and there wouldn't be a need to remember the tracing type. 

If I'm missing something, please help point it out.

Comment (yurishkuro) [reply src/jaegertracing/SpanContext.h]: The code snippet you showed is exactly how you do it. The tracer is created once at application start-up, and at that time you can replace the codecs. After that no more "dynamic" configuration is needed, nor does the span need to remember which codec is used.

Comment (kunalekawde) [reply src/jaegertracing/SpanContext.h]: oh ok, which means that application has to be aware of what type tracing type has to supported explicitly which can be either Uber or B3 mutually exclusively.

Comment (yurishkuro) [reply src/jaegertracing/SpanContext.h]: Yes, but the type of context used on the wire is usually the same for the whole site, not varied per application or per request.

It is theoretically possible to write a composite codec combined from both Uber and B3 headers, if it's required for backwards compatibility. But a single format for the whole system is highly recommended.

Comment (kunalekawde) [reply src/jaegertracing/SpanContext.h]: Sure thanks, I had made the changes to handle kind of dynamic selection expecting either Uber or B3 headers at runtime at the node. An API getTracerType checks for incoming headers and uses particular tracer type - Uber/B3.

Comment (yurishkuro) [reply src/jaegertracing/SpanContext.h]: Sorry, but I don't think we want to support this "dynamic" behavior. None of the other Jaeger clients do it this way, and I have not seen anyone ever asking for this. The codec is configured when creating a tracer, it does not change after that. A composite codec can be implemented ([example in java](https://github.com/jaegertracing/jaeger-client-java/blob/dfb9d8b9dc849f2e6f880b44b26fb77dfc894a02/jaeger-core/src/main/java/io/jaegertracing/internal/propagation/CompositeCodec.java)), which handles more than one encoding, but it is also doing them _always_, not dynamically. As a result, there's never a need to store "type" in the span context.

I don't know why you're trying to reinvent the wheel. I pointed you to several examples how codecs are done in Jaeger clients in other languages, it's an established pattern.

Comment (kunalekawde) [reply src/jaegertracing/SpanContext.h]: Not being certain of the ingress nodes and its supported types(e.g. Istio currently supports only B3) we thought being flexible to handle any type based on incoming headers would be good. 
Yes, as per the case you mentioned of static provisioning this PR would need more enhancement. Thanks for all the help and references. I shall probably create a new one and close this.

**Comment by kunalekawde** [comment]
closing this PR to have static provisioning of tracing types for which PR needs to be enhanced.
